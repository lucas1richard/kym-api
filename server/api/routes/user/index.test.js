const { expect, use } = require('chai');
const { jestSnapshotPlugin } = require('mocha-chai-jest-snapshot');

describe('user routes', () => {
  use(jestSnapshotPlugin());
  const routePrefix = '/api/user';

  describe(`GET ${routePrefix}/v1`, () => {
    const route = `${routePrefix}/v1`;
    before(globals.seedTestData);
    after(globals.destroyAllHook);

    it('gets the user', async () => {
      const res = await globals.agent.getWithToken(route);
      expect(res.statusCode).equal(200);
      expect(res.body).toMatchSnapshot();
    });
  });

  describe(`PUT ${routePrefix}/v1`, () => {
    const route = `${routePrefix}/v1`;
    before(globals.seedTestData);
    after(globals.destroyAllHook);

    it('updates the user', async () => {
      const res = await globals.agent.putWithToken(route).send({ firstname: 'Mockname' });
      expect(res.statusCode).equal(201);
      expect(res.body.firstname).equal('Mockname');
    });
    it('validats the req', async () => {
      const res = await globals.agent.putWithToken(route).send({ firstname: 1 });
      expect(res.statusCode).equal(400);
    });
  });

  describe(`POST ${routePrefix}/measurements/v1`, () => {
    const route = `${routePrefix}/measurements/v1`;
    before(globals.seedTestData);
    after(globals.destroyAllHook);

    const measurementBody = {
      date: '2022-09-01',
      weight: '84',
      age: '33',
      gender: 'MALE',
      goal: 'LOSE_FAT',
      height: '186',
      lifestyle: 'NORMAL',
      units: 'METRIC',
    };

    it('creates measurements for a user for the current date', async () => {
      const res = await globals.agent.postWithToken(route).send({
        ...measurementBody,
        date: undefined,
      });
      expect(res.statusCode).equal(201);
      expect(res.body.weight).equal('84');
    });

    it('creates measurements for a user', async () => {
      const res = await globals.agent.postWithToken(route).send(measurementBody);
      expect(res.statusCode).equal(201);
      expect(res.body.weight).equal('84');
    });

    it('modifies existing measurements for a user', async () => {
      const res = await globals.agent.postWithToken(route).send({
        ...measurementBody,
        weight: '83',
      });
      expect(res.statusCode).equal(201);
      expect(res.body.weight).equal('83');
    });
    it('validates the req', async () => {
      const res = await globals.agent.postWithToken(route).send({});
      expect(res.statusCode).equal(400);
    });
  });

  describe(`PUT ${routePrefix}/measurements/v1`, () => {
    const route = `${routePrefix}/measurements/v1`;
    before(globals.seedTestData);
    after(globals.destroyAllHook);

    const measurementBody = {
      id: 1,
      date: '2022-09-01',
      weight: '84',
      age: '33',
      gender: 'MALE',
      goal: 'LOSE_FAT',
      height: '186',
      lifestyle: 'NORMAL',
      units: 'METRIC',
    };

    it('creates measurements for a user for the current date', async () => {
      const res = await globals.agent.putWithToken(route).send({
        ...measurementBody,
      });
      expect(res.statusCode).equal(201);
      expect(res.body.weight).equal('84');
    });

    it('ensures the measurment exists before updating', async () => {
      const res = await globals.agent.putWithToken(route).send({
        ...measurementBody,
        id: 10000000,
      });
      expect(res.statusCode).equal(400);
    });
  });

  describe(`DELETE ${routePrefix}/measurements/v1`, () => {
    const route = `${routePrefix}/measurements/v1`;
    before(globals.seedTestData);
    after(globals.destroyAllHook);

    it('deletes a measurement', async () => {
      const res = await globals.agent.deleteWithToken(route).send({
        id: 1,
      });
      expect(res.statusCode).equal(204);
    });

    it('validates the req', async () => {
      const res = await globals.agent.deleteWithToken(route).send({});
      expect(res.statusCode).equal(400);
    });
  });
});
