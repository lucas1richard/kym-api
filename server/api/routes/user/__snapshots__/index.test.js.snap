// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user routes GET /api/user/v1 gets the user 1`] = `
Object {
  "birthdate": "1989-01-01",
  "email": "test@test.test",
  "firstname": "Test",
  "lastname": "Test",
  "mealGoals": Array [
    Object {
      "goals": Object {
        "REST": Array [
          Object {
            "carbs": 29.3,
            "fat": 13.1,
            "protein": 33.3,
          },
          Object {
            "carbs": 29.3,
            "fat": 13.1,
            "protein": 33.3,
          },
          Object {
            "carbs": 77.8,
            "fat": 34.7,
            "protein": 88.2,
          },
          Object {
            "carbs": 0,
            "fat": 0,
            "protein": 0,
          },
          Object {
            "carbs": 64.6,
            "fat": 28.8,
            "protein": 73.2,
          },
          Object {
            "carbs": 0,
            "fat": 0,
            "protein": 0,
          },
        ],
        "TRAIN": Array [
          Object {
            "carbs": 79.5,
            "fat": 29.9,
            "protein": 63.3,
          },
          Object {
            "carbs": 0,
            "fat": 0,
            "protein": 0,
          },
          Object {
            "carbs": 119.2,
            "fat": 44.8,
            "protein": 95,
          },
          Object {
            "carbs": 0,
            "fat": 0,
            "protein": 0,
          },
          Object {
            "carbs": 87.4,
            "fat": 32.8,
            "protein": 69.7,
          },
          Object {
            "carbs": 0,
            "fat": 0,
            "protein": 0,
          },
        ],
      },
      "id": 1,
      "user_uuid": "60f24265-ef45-4891-b4ef-e8fad4435308",
    },
  ],
  "preferredlocale": null,
  "programs": Array [
    Object {
      "enddate": "2022-10-23",
      "endgoal": "180",
      "endweight": null,
      "id": 1,
      "result": "TBD",
      "startdate": "2022-09-18",
      "startweight": "175",
      "status": "IN_PROGRESS",
      "user_uuid": "60f24265-ef45-4891-b4ef-e8fad4435308",
    },
  ],
  "userMeasurements": Array [
    Object {
      "age": "28",
      "bmrBodyFat": 1900,
      "bmrTraditional": 1876,
      "bodyfat": "16",
      "date": "2018-01-01T00:00:00.000Z",
      "gender": "MALE",
      "goal": "LOSE_FAT",
      "height": "186",
      "id": 1,
      "lifestyle": "NORMAL",
      "units": "METRIC",
      "user_uuid": "60f24265-ef45-4891-b4ef-e8fad4435308",
      "weight": "86",
    },
  ],
  "username": "",
  "uuid": "60f24265-ef45-4891-b4ef-e8fad4435308",
}
`;
